core/trino-main/src/main/java/io/trino/operator/aggregation/OrderedWindowAccumulator.java

```diff
@@ -0,0 +1,145 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.aggregation;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.operator.PagesIndex;
+import io.trino.operator.window.PagesWindowIndex;
+import io.trino.spi.PageBuilder;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.ValueBlock;
+import io.trino.spi.connector.SortOrder;
+import io.trino.spi.function.WindowAccumulator;
+import io.trino.spi.function.WindowIndex;
+import io.trino.spi.type.Type;
+
+import java.util.List;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class OrderedWindowAccumulator
+        implements WindowAccumulator
+{
+    private final PagesIndex.Factory pagesIndexFactory;
+    private final PagesIndex pagesIndex;
+    private final List<Type> argumentTypes;
+    private final List<Integer> argumentChannels;
+    private final List<Integer> sortChannels;
+    private final List<SortOrder> sortOrders;
+
+    private WindowAccumulator delegate;
+    private final WindowAccumulator initialDelegate;
+
+    private PageBuilder pageBuilder;
+    private boolean pagesIndexSorted;
+
+    public OrderedWindowAccumulator(
+            PagesIndex.Factory pagesIndexFactory,
+            WindowAccumulator delegate,
+            List<Type> argumentTypes,
+            List<Integer> argumentChannels,
+            List<SortOrder> sortOrders)
+    {
+        this(pagesIndexFactory, pagesIndexFactory.newPagesIndex(argumentTypes, 10_000), delegate, argumentTypes, argumentChannels, sortOrders);
+    }
+
+    private OrderedWindowAccumulator(
+            PagesIndex.Factory pagesIndexFactory,
+            PagesIndex pagesIndex,
+            WindowAccumulator delegate,
+            List<Type> argumentTypes,
+            List<Integer> argumentChannels,
+            List<SortOrder> sortOrders)
+    {
+        this.pagesIndexFactory = requireNonNull(pagesIndexFactory, "pagesIndexFactory is null");
+        this.pagesIndex = requireNonNull(pagesIndex, "pagesIndex is null");
+        this.argumentTypes = requireNonNull(argumentTypes, "argumentTypes is null");
+        this.argumentChannels = ImmutableList.copyOf(argumentChannels);
+        this.sortOrders = ImmutableList.copyOf(sortOrders);
+        this.sortChannels = IntStream.range(argumentTypes.size() - sortOrders.size(), argumentChannels.size())
+                .boxed()
+                .toList();
+
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.initialDelegate = delegate.copy();
+
+        resetPageBuilder();
+    }
+
+    private void resetPageBuilder()
+    {
+        pageBuilder = new PageBuilder(argumentTypes);
```

Comment:
why new PageBuilder instead of pageBuilder.reset()?

Reply:
did not know about that method :)